presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java

```diff
@@ -123,80 +125,86 @@ private PlanNodeStatsEstimate filterForFalseExpression()
                                             .setDistinctValuesCount(0.0)
                                             .setNullsFraction(NaN).build()));
 
-            return falseStatsBuilder.setOutputRowCount(0.0).build();
+            return Optional.of(falseStatsBuilder.setOutputRowCount(0.0).build());
         }
 
         @Override
-        protected PlanNodeStatsEstimate visitNotExpression(NotExpression node, Void context)
+        protected Optional<PlanNodeStatsEstimate> visitNotExpression(NotExpression node, Void context)
         {
-            return differenceInStats(input, process(node.getValue()));
+            return process(node.getValue()).map(childStats -> differenceInStats(input, childStats));
         }
 
         @Override
-        protected PlanNodeStatsEstimate visitLogicalBinaryExpression(LogicalBinaryExpression node, Void context)
+        protected Optional<PlanNodeStatsEstimate> visitLogicalBinaryExpression(LogicalBinaryExpression node, Void context)
         {
-            PlanNodeStatsEstimate leftStats = process(node.getLeft());
-            PlanNodeStatsEstimate andStats = new FilterExpressionStatsCalculatingVisitor(leftStats, session, types).process(node.getRight());
-
+            Optional<PlanNodeStatsEstimate> leftStats = process(node.getLeft());
+            Optional<PlanNodeStatsEstimate> andStats = new FilterExpressionStatsCalculatingVisitor(leftStats.orElse(filterForUnknownExpression()), session, types).process(node.getRight());
```

Comment:
Yeah, you're right.  
if you don't mind some extra computation you could do something like below.   
```
leftStats = process(node.getLeft())
rightStats = process(node.getRight());
if(leftStats.isPresent() && rightStats.isPresent()) {
    andStats = FilterExpressionStatsCalculator(leftStats).process(node.getRight())
} 
else if(leftStats.isPresent() && !rightStats.isPresent()) {
   andStats = leftStats.map(FilterStatsCalculator::filterStatsForUnknownExpression);
}
else if(rightStats.isPresent() && !leftStats.isPresent()) {
   andStats = rightStats.map(FilterStatsCalculator::filterStatsForUnknownExpression);
} 
else {
   andStats = Optional.empty()
}
```
Clunky as that is I think it's easier to follow because it's more explicit.  Or maybe it will inspire you with some other idea.